### PR TITLE
Inspect: Allow showing data without transformations and field config applied

### LINKED ISSUE #24285: PanelInspect: allow showing data without transformations and field config applied
Panel inspect Data table now shows the data _with transformations and field configuration applied_. At the same time the CSV downloaded shows the actual raw data.

As discussed with @torkelo, @sarlinska and @jessover9000  It would be good to:
1. Add a possibility to select whether or not transformations should be applied
2. Add a possibility to select whether or not field config should be applied

### PR DESCRIPTION
**Pull request by torkelo**
Fixes https://github.com/grafana/grafana/issues/24285

Trying to solve the problem (start) of inspect showing table custom display options (cell display modes) & the transformed data, with no way to disable the transform an look at raw data. 

* Updates PanelQueryRunnner getData to not run applyFieldOverrides if transform = false, so transform = false 
means get give me raw un transformed & un processed (field config) data.  Do we ever want not transformed but with field config applied? hm..probably some case? 

* Inspect is currently calling applyFieldOverrides itself but with no field config defaults or overrides. So this means inspect will have nothing of the panel config defaults or overrides applied (like units etc). Good / bad? 

For the use case of viewing and exporting the transformed data, I think we need an option like a toggle. "With panel transforms" or something so a user can decide if they want the panel transforms or not. Hm... and that makes the change to PanelQueryRunnner a bit wrong I guess, as then we want to get transformed data & have some control over field config. 

maybe getData needs something like getData({transform: true/false, fieldConfigRegistry: none|standard|all)

### CODE REVIEW

**Comment by dprokop** [comment]
@torkelo I like the idea of improving the API of getData. Probably I wouldn't bother with `fieldConfigRegistry` option thought, and just keep it simple:

`getData({transform: boolean, applyFieldConfig: boolean});`

In general I think whatever data comes into the InspectDataTab component this is what should get rendered, without any additional processing (apart from transformations that you can apply in the inspector).

**Comment by torkelo** [comment]
> getData({transform: boolean, applyFieldConfig: boolean});

:+1:

**Comment by ryantxu** [comment]
Does it make more sense to explicitly ask for the raw data?  And only show the option when transformations exist?

**Comment by dprokop** [comment]
@ryantxu what do you mean? In the implementation transformations switch is only available if there are transformations configured in the panel. Otherwise it's only possible to apply field config:

image

vs

image

**Comment by ryantxu** [comment]
Yes, I would flip it -- I expect to see the same data as the panel (with transformations and field config) unless I explicitly want the raw data from the query

**Comment by dprokop** [comment]
ah, I see. WDYT @torkelo ?

File: public/app/features/dashboard/components/Inspector/InspectDataTab.tsx [hunk 1]
```diff
@@ -67,25 +84,34 @@ export class InspectDataTab extends PureComponent<Props, State> {
 
     const currentTransform = transformationOptions.find(item => item.value === transformId);
 
+    console.log(data);
```
Comment (dprokop) [inline public/app/features/dashboard/components/Inspector/InspectDataTab.tsx:87]:
```suggestion
```

File: public/app/features/dashboard/components/Inspector/InspectDataTab.tsx [hunk 2]
```diff
@@ -67,25 +84,34 @@ export class InspectDataTab extends PureComponent<Props, State> {
...
     if (currentTransform && currentTransform.transformer.id !== DataTransformerID.noop) {
       return transformDataFrame([currentTransform.transformer], data);
     }
     return data;
   }
 
   getProcessedData(): DataFrame[] {
+    if (this.state.transformId === DataTransformerID.noop) {
+      console.log(this.props.data);
```
Comment (dprokop) [inline public/app/features/dashboard/components/Inspector/InspectDataTab.tsx:96]:
```suggestion
```

**Comment by dprokop** [comment]
@torkelo @ryantxu changed so that transformations are applied by default. Field config is not.

**Comment by torkelo** [comment]
First time I go into inspect I get a lot of decimals, are we skipping applyFieldOverrides entirely? 

Screenshot from 2020-05-11 11-50-24

**Comment by dprokop** [comment]
> First time I go into inspect I get a lot of decimals, are we skipping applyFieldOverrides entirely?

Raw is raw :)

**Comment by torkelo** [comment]
The decimals disappear if I apply a transformation or check the field config options

**Comment by dprokop** [comment]
> The decimals disappear if I apply a transformation or check the field config options

@torkelo fixed this. Sorry, applied fix in the table panel and haven't removed this bug. Now it's fixed in e9fdadd5ff2cc5760e6374234b4958e47d296b79

**Comment by torkelo** [comment]
I think the "apply panel transformations" should be an option in the transform selector, and the "apply field configuration option" switch could be a switch next to the transform selector? Maybe we can wait with this switch and always apply field config options. 

Also the time formatting is lost (we need to pass this through applyFieldOverrides function to get default formatting). 

We have another option and that all the options (transform) and frame select into the data options dropdown. Just feels strange now UX wise to have options in two places, one in a collapsed row below and one above. 

The CSV download is not affected by any of the field config options, always get just raw data (no time formatting, or decimals).  

I now there is a lot to think about and potentially change / fix here. Question is what should we do in this PR and for 7.0

**Comment by dprokop** [comment]
@torkelo for readability purposes I'm applying the time formatting now in the table. Question - do you think this formatting should reflect the dashboard time zone?

**Comment by dprokop** [comment]
> I now there is a lot to think about and potentially change / fix here. Question is what should we do in this PR and for 7.0

IMO we should proceed with what's done here, and apply UX changes and the CSV download in 7.1.

File: public/app/features/dashboard/components/Inspector/InspectDataTab.tsx [hunk 3]
```diff
@@ -73,19 +94,43 @@ export class InspectDataTab extends PureComponent<Props, State> {
...
+          field.display = getDisplayProcessor({
+            field,
+            theme: config.theme,
+          });
+        }
+      }
+    }
+
+    return data;
+  };
```
Comment (torkelo) [inline public/app/features/dashboard/components/Inspector/InspectDataTab.tsx:110]: Think I understand the reasons, but this feels wrong, this is handled in applyFieldOverrides, should be able to use that somehow. And if no field config options are applied we only show raw data.

Comment (dprokop) [reply public/app/features/dashboard/components/Inspector/InspectDataTab.tsx:110]: yeah, i'm about to change this, as I hate ^^ approach

File: docs/sources/panels/inspect-panel.md [hunk 4]
```diff
@@ -42,14 +42,19 @@ The panel inspector pane opens on the right side of the screen.
 
 ### Inspect raw query results
 
-View raw query results in a table. This is the data returned by the query with transformations applied and before the panel applies field configurations or overrides.
+View raw query results in a table. This is the data returned by the query with transformations applied and before the panel applies field options or options overrides.
```
Comment (oddlittlebird) [inline docs/sources/panels/inspect-panel.md:45]:
```suggestion
View raw query results in a table. This is the data returned by the query with transformations applied and before the panel applies field options or field option overrides. 
```

File: docs/sources/panels/inspect-panel.md [hunk 5]
```diff
@@ -42,14 +42,19 @@ The panel inspector pane opens on the right side of the screen.
...
+View raw query results in a table. This is the data returned by the query with transformations applied and before the panel applies field options or options overrides. 
 
 1. Open the panel inspector and then click the **Data** tab or in the panel menu click **Inspect > Data**.
 
 2. If your panel contains multiple queries or queries multiple nodes, then you have additional options.
 * **Select result -** Choose which result set data you want to view.
 * **Transform data**
   * **Join by time -** View raw data from all your queries at once, one result set per column. Click a column heading to reorder the data.
+  
+  View raw query results in a table with field options and options overrides applied:
```
Comment (oddlittlebird) [inline docs/sources/panels/inspect-panel.md:54]: Please use lazy numbering.

**Review by oddlittlebird — CHANGES_REQUESTED** [review]

File: public/app/features/dashboard/components/Inspector/InspectDataTab.tsx (hunk 3 above)
Comment (dprokop) [reply public/app/features/dashboard/components/Inspector/InspectDataTab.tsx]: @torkelo this is now updated and ready for another take from you :)

File: docs/sources/panels/inspect-panel.md (hunk 5 above)
Comment (dprokop) [reply docs/sources/panels/inspect-panel.md:54]: fixed

File: public/app/features/dashboard/state/PanelQueryRunner.ts [hunk 6]
```diff
@@ -51,6 +51,15 @@ function getNextRequestId() {
   return 'Q' + counter++;
 }
 
+export interface GetDataOptions {
+  transform?: boolean;
+  applyFieldConfig?: boolean;
```
Comment (torkelo) [inline public/app/features/dashboard/state/PanelQueryRunner.ts:56]: nitpick: always struggle a bit with verb booleans. 

withTransforms: boolean, 
withFieldConfig: boolean

?

File: public/app/features/dashboard/state/PanelQueryRunner.ts [hunk 7]
```diff
@@ -66,37 +75,44 @@ export class PanelQueryRunner {
...
    * Returns an observable that subscribes to the shared multi-cast subject (that reply last result).
    */
-  getData(transform = true): Observable<PanelData> {
+  getData(options?: GetDataOptions): Observable<PanelData> {
+    const { transform, applyFieldConfig } = options
+      ? {
+          ...DEFAULT_GET_DATA_OPTIONS,
+          ...options,
+        }
+      : DEFAULT_GET_DATA_OPTIONS;
```
Comment (torkelo) [inline public/app/features/dashboard/state/PanelQueryRunner.ts:84]: can't we do getData(options? = DEFAULT_GET_DATA_OPTIONS), might be good require options & the bool flags, to make this more explicit so we do not have more bugs with wrong data.

File: public/app/plugins/datasource/dashboard/runSharedRequest.ts [hunk 8]
```diff
@@ -35,7 +35,7 @@ export function runSharedRequest(options: QueryRunnerOptions): Observable<PanelD
     }
 
     const listenToRunner = listenToPanel.getQueryRunner();
-    const subscription = listenToRunner.getData(false).subscribe({
+    const subscription = listenToRunner.getData({ transform: false }).subscribe({
```
Comment (torkelo) [inline public/app/plugins/datasource/dashboard/runSharedRequest.ts:38]: I think this should be false for field config as well

**Review by torkelo — COMMENTED** [review]
Sorry, still not 100% happy with this, especially UX. 

Don't like that these options are not part of the main options row, away from the transform & frame select options. make it very inconsistent. 

Screen Shot 2020-05-12 at 13 42 38

And looks strange when no options 
Screen Shot 2020-05-12 at 13 48 21

And if feels like the "Panel transforms" options should be an option in the transform select, not a standalone switch.

**Comment by dprokop** [comment]
>Sorry, still not 100% happy with this, especially UX.

Updated according to what we discussed

File: public/app/features/dashboard/components/Inspector/PanelInspector.tsx [hunk 9]
```diff
@@ -99,11 +108,12 @@ export class PanelInspectorUnconnected extends PureComponent<Props, State> {
    */
   init() {
     const { plugin, panel } = this.props;
+    const { withTransforms, withFieldConfig } = this.state;
 
     if (plugin && !plugin.meta.skipDataQuery) {
       this.querySubscription = panel
         .getQueryRunner()
-        .getData()
+        .getData({ withTransforms, withFieldConfig })
```
Comment (torkelo) [inline public/app/features/dashboard/components/Inspector/PanelInspector.tsx:116]: This function is called many times, need to check if query subscription exists an unsubscribe before subscribing again

**Review by torkelo — COMMENTED** [review]
found one more thing with the subscription

**Review by dprokop — APPROVED** [review]

**Pull request merged**